enterprise/server/api/api_server.go

```diff
@@ -147,6 +148,42 @@ func (s *APIServer) GetTarget(ctx context.Context, req *apipb.GetTargetRequest)
 	}, nil
 }
 
+func (s *APIServer) getActionsFromCache(ctx context.Context, req *apipb.GetActionRequest) ([]*apipb.Action, error) {
+	// This limit is a safeguard against buffering too many results in memory.
+	// We expect to hit this rarely (if ever) in production usage.
+	const limit = 100_000
+	if !api_config.CacheEnabled() || s.env.GetMetricsCollector() == nil {
+		return nil, nil
+	}
+	iid := req.GetSelector().GetInvocationId()
+	serializedResults, err := s.env.GetMetricsCollector().ListRange(ctx, api_common.ActionsKey(iid), 0, limit-1)
+	if err == nil {
+		return nil, err
+	}
+	dec, err := zstd.NewReader(nil)
+	if err != nil {
+		return nil, err
+	}
+
+	actions := make([]*apipb.Action, 0)
+	a := &apipb.Action{}
+	buf := make([]byte, 100)
+
+	for _, compressed := range serializedResults {
+		raw, err := dec.DecodeAll([]byte(compressed), buf[:0])
```

Comment:
might make sense to do `compression.DecompressZstd` here so that you go through the DecoderPool